enhancements/machine-api/aws-efa-networking.md

```diff
@@ -0,0 +1,357 @@
+any user automating changes to such security groups is already exposing themself to potential compatibility issues with
+future versions of OCP.
+
+#### Users will have to configure Huge Pages
+
+Huge Pages is a prerequisite requirement for deploying MPI workloads which leverage the EFA network interface.
+It is expected that users will be responsible for configuring this within their OpenShift clusters post installation.
+
+We [already document](https://docs.openshift.com/container-platform/4.9/post_installation_configuration/node-tasks.html#configuring-huge-pages_post-install-node-tasks)
+how to configure Huge Pages within OpenShift clusters as a day 2 operation and it is expected that users of this
+feature will follow the instructions linked above.
+
+For completeness, an example configuration is included below:
+
+```yaml
+apiVersion: tuned.openshift.io/v1
+kind: Tuned
+metadata:
+  name: hugepages
+  namespace: openshift-cluster-node-tuning-operator
+spec:
+  profile:
+  - data: |
+      [main]
+      summary=Boot time configuration for hugepages
+      include=openshift-node
+      [bootloader]
+      cmdline_openshift_node_hugepages=hugepagesz=2M hugepages=10256M
+    name: openshift-node-hugepages
+
+  recommend:
+  - machineConfigLabels:
+      machineconfiguration.openshift.io/role: "worker"
+    priority: 30
+    profile: openshift-node-hugepages
+```
+
+Note, as the Huge Pages configuration is tuned to the size of the instance types leveraged within the cluster,
+it is hard to predict an appropriate default. Due to this limitation, we leave this configuration as an exercise
+for the end user.
+
+A `2M` page size is the minimum requirement for the MPI operator.
+The number of Huge Pages must fit within the hosts memory.
+
+#### Using EFA requires additional software
+
+The EFA network interface types require special drivers to enable them to be used.
+Fortunately, the RHEL kernel already includes the relevant kernel modules to allow the adapter to be used transparently
+as a standard network adapter within user space. This means that no additional configuration is required within RHCOS
+to allow EFA instances to be used as regular OpenShift nodes, existing networking is unaffected.
+
+To take advantage of the EFA performance benefits, additional libfabric software must be deployed.
+This software is currently [distributed by AWS](https://github.com/aws-samples/aws-efa-eks/blob/main/manifest/efa-k8s-device-plugin.yml)
+and it will be up to the end user to deploy this software within their clusters.
+Red Hat will not initially support the libfabric driver.
+
+Once the end user has correctly configured the software prerequisites, the Node Feature Discovery operator will
+configure the Node status to show the EFA interface as an allocatble resource.
+This then allows the MPI operator to start scheduling MPI workloads onto the Node.
```

Comment:
is there any delay in the timing between the instance becoming a node and the time it takes to advertise the EFA capacity? (is this even a thing)

i'm wondering if the autoscaler will need to account for the time it takes to deploy the EFA driver (in a similar manner as we do for the GPU drivers today) ?

Reply:
Yes I think there is. I believe this will behave very similarly to how GPU autoscaling functions. Is the GPU also detected by the node feature discovery operator?

Reply:
yeah, i believe NFD is what does the detection. i was more curious about what we might need to do to support this on the autoscaler side and it sounds like we will probably have to add some special detection logic once we figure out how these nodes advertise their EFA status.

Reply:
Again, we also need Topolo awareness into this picture, one thing is NFD labeling EFA as available and other thing is to have a cluster of nodes in close proximity to say that MPI will be boosted by a fast network interface. (I have tested the MPI operator and it can go as wrong as allocating nodes on separate regions)

Reply:
i would expect this to be defined in the machineset, and that will instruct the autoscaler on where to place things. for example:

user defines 1 machineset for autoscaling with efa enabled. the efa-enabled machineset will have the placement information in its provider spec. when a workload is scheduled to that machineset (efa-enabled), and there are no machines available, the autoscaler will scale the replicas on that machineset and the underlying aws infrastructure will be ensuring that those machines are placed in the same rack.

now, this becomes more complex if the user creates 2 machinesets that both have efa enabled. we can't have a workload designed to run at low latency being spread across the 2 machinesets (even though both are configured for efa and proper placement) unless we can guarantee that those 2 machinesets are also in the same rack. in this case, i would expect the user to have label selectors for the 2 machinesets and control their workloads by using that selector.

in the end, the autoscaler won't be doing any topology-aware calculcations that are specific to the placement, but our usage of machinesets will effectively control how the autoscaler can scale out. i think the most difficult cases will be where our users want to spread a workload that might go beyond the maximum number of machines that can exist in a local (same rack) placement group. i'm not sure how we could solve for that. but for workloads that are smaller, i think we can solve the issue using the machinesets and selector labels (applied by the user).